channels/tsmf/client/tsmf_codec.c

```diff
@@ -411,6 +524,114 @@ static UINT32 tsmf_codec_parse_VIDEOINFOHEADER(TS_AM_MEDIA_TYPE* mediatype, wStr
+}
+
+static UINT32 tsmf_codec_parse_HEAACWAVEINFOHEADER(TS_AM_MEDIA_TYPE* mediatype, wStream* s)
+{
+	/*
+		typedef struct heaacwaveformat_tag {
+		  HEAACWAVEINFO wfInfo;
+		  BYTE          pbAudioSpecificConfig[1];
+		} HEAACWAVEFORMAT, *PHEAACWAVEFORMAT;
+		
+		typedef struct heaacwaveinfo_tag {
+		  WAVEFORMATEX wfx;
+		  WORD         wPayloadType;
+		  WORD         wAudioProfileLevelIndication;
+		  WORD         wStructType;
+		  WORD         wReserved1;
+		  DWORD        dwReserved2;
+		} HEAACWAVEINFO, *PHEAACWAVEINFO;
+	*/
+	UINT16 wPayloadType;
+	UINT16 wAudioProfileLevelIndication;
+	UINT16 wStructType;
+
+	Stream_Read_UINT16(s, wPayloadType);
+	Stream_Read_UINT16(s, wAudioProfileLevelIndication);
+	Stream_Read_UINT16(s, wStructType);
+	Stream_Seek(s, 6);
+	mediatype->ExtraDataSize -= 12;
+	if(mediatype->ExtraDataSize >= 2){
+		BYTE *pbAudioSpecificConfig = Stream_Pointer(s);
+		int audioObjectType = (pbAudioSpecificConfig[0] & 0xF8) >> 3; // 取前5bit
+		int samplingFrequencyIndex = ((pbAudioSpecificConfig[0] & 0x7) << 1) | (pbAudioSpecificConfig[1] >> 7);//取后4bit，结果为3
+		int channelConfiguration = (pbAudioSpecificConfig[1] >> 3) & 0x0F; // 取后4bit
+		int frameLengthFlag = (pbAudioSpecificConfig[1] >> 2) & 0x01;// 取后1bit
+		int dependsOnCoreCoder = (pbAudioSpecificConfig[1] >> 1) & 0x01; // 取后1bit
+		int extensionFlag = pbAudioSpecificConfig[1] & 0x01; // 最后1bit
+		DEBUG_TSMF(" audioObjectType:0x%"PRIx32, audioObjectType);
+		DEBUG_TSMF(" wPayloadType:0x%"PRIx16" wAudioProfileLevelIndication:0x%"PRIx16" wStructType:0x%"PRIx16" audioObjectType:0x%"PRIx32" samplingFrequencyIndex:0x%"PRIx32" channelConfiguration:0x%"PRIx32" frameLengthFlag:0x%"PRIx32" dependsOnCoreCoder:0x%"PRIx32" extensionFlag:0x%"PRIx32, wPayloadType, wAudioProfileLevelIndication, wStructType, audioObjectType, samplingFrequencyIndex, channelConfiguration, frameLengthFlag, dependsOnCoreCoder, extensionFlag);
+		/* ISO/IEC 14496-3
+		Check for the presence of SBR data in the bitstream payload prior to decoding. If no SBR data is found, the output sample rate is equal to that signaled as samplingFrequency in the AudioSpecificConfig(). If SBR data is found the output sample rate is twice that signaled as samplingFrequency in the AudioSpecificConfig
+		*/
+		/* https://technet.microsoft.com/zh-cn/dd742784(v=vs.90).aspx
+		The length of the AudioSpecificConfig() data is 2 bytes for AAC-LC or HE-AAC with implicit signaling of SBR/PS. It is more than 2 bytes for HE-AAC with explicit signaling of SBR/PS. The value of audioObjectType as defined in AudioSpecificConfig() must be 2, indicating AAC-LC. The value of extensionAudioObjectType must be 5 for SBR or 29 for PS.
+		*/
+
+		if(audioObjectType!=2){
+			WLog_ERR(TAG, "wrong audioObjectType!");
+		}else{
+			int bits_to_decode = (mediatype->ExtraDataSize-2) * 8;
+			pbAudioSpecificConfig = Stream_Pointer(s) + 2;
+
+			DEBUG_TSMF("SBR: bits_to_decode() %d", bits_to_decode);
+			if(bits_to_decode >= 16){ // left bits is sbr?!
```

Comment:
For this big statement with stairs, may I suggest to do:
```
do 
{
    if (bits_to_decode < 16);
         break;
    .....
    if(syncExtensionType != 0x2b7)
         break;
    .....
    if(extensionAudioObjectType != 0x05)
          break;

   .... (and do on)....

} while(0);
```
The code will look nicer with less stairs of if().

Plus you should definitely look at the FreeRDP code style (curly bracket are supposed to be on a single line, binary operators are supposed to have spaces around them, things like that).